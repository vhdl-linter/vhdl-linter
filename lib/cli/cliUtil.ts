import { readdirSync } from "fs";
import { cwd } from "process";
import { DiagnosticSeverity } from "vscode-languageserver";
import { OIRange } from "../parser/objects";
import { joinURL } from "../projectParser";
import { OIDiagnostic } from "../vhdlLinter";

export interface MessageWrapper {
  file: string,
  messages: (OIDiagnostic | { message: string })[]
}
function isOIDiagnostic(obj: unknown): obj is OIDiagnostic {
  if ((obj as OIDiagnostic).range instanceof OIRange) {
    return true;
  }
  return false;
}
export function readDirPath(path: URL) {
  return readdirSync(path).map(file => joinURL(path, file));
}
function getMessageColor(message: OIDiagnostic | { message: string }) {
  if (isOIDiagnostic(message) && message.severity === DiagnosticSeverity.Error) {
    return '\u001b[31m';
  } else if (isOIDiagnostic(message) && message.severity === DiagnosticSeverity.Warning) {
    return '\u001b[33m';
  }
  return '\u001b[34m';
}
export function prettyPrintMessages(messages: MessageWrapper[]) {
  return messages.map(message => {
    const filename = message.file.replace(cwd(), '');
    return message.messages.slice(0, 5).map((innerMessage) => {
      const messageText = `${getMessageColor(innerMessage)}${innerMessage.message}\u001b[0m`;
      if (isOIDiagnostic(innerMessage)) {
        return `${filename}:${innerMessage.range.start.line + 1} (r: ${innerMessage.range.start.line}:${innerMessage.range.start.character} - ${innerMessage.range.end.line}:${innerMessage.range.end.character})\n  ${messageText}`; // lines are 0 based in OI
      }
      return `${filename}\n  ${messageText}`;
    }).join('\n') + (message.messages.length > 5 ? `\n\u001b[31m ... and ${message.messages.length - 5} more\u001b[0m` : '');
  }).join('\n');
}
