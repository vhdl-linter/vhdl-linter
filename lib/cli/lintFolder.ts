import { lstatSync } from "fs";
import { ProjectParser, vhdlGlob } from "../projectParser";
import { VhdlLinter } from "../vhdlLinter";
import { readFileSyncNorm } from "./readFileSyncNorm";
import { MessageWrapper, prettyPrintMessages, readDirPath } from "./cliUtil";
import { OIRange } from "../parser/objects";
import { minimatch } from "minimatch";
import { basename, dirname } from "path";
import { fileURLToPath } from "url";
import { execSync } from "child_process";
import { platform } from "process";

let gitExists = false;
if (platform === 'win32') {
  try {
    execSync(`git --version`);
    gitExists = true;
  } catch (e) {
    gitExists = false;
  }
} else {
  try {
    execSync(`command -v git`);
    gitExists = true;
  } catch (e) {
    gitExists = false;
  }
}
function isGitIgnored(path: string) {
  if (path.endsWith('.git')) {
    return true;
  }
  if (gitExists) {
    // git check-ignore will fail if the file is *NOT* ignored or when we are not in a git repository
    try {
      execSync(`git check-ignore ${path}`, {cwd: dirname(path)});
      // is a git repo and the file is ignored...
      return true;
    } catch (e) {
      return false;
    }
  } else {
    return false;
  }
}
function ignore(excludePatterns: string[], relativePath: string, absolutePath: string) {
  return excludePatterns.some(glob => minimatch(relativePath, glob)) || isGitIgnored(absolutePath);
}

export async function lintFolder(path: URL, errorExpected: boolean, printMessages: boolean, excludePatterns: string[], projectParser: ProjectParser): Promise<MessageWrapper[]> {
  const result: MessageWrapper[] = [];
  const rootPath = fileURLToPath(projectParser.workspaces[0]!);
  for (const subPath of readDirPath(path)) {
    // in this case we only ever have one workspace
    const relativePath = fileURLToPath(subPath).replace(rootPath, '').substring(1);
    // ignore excluded folders & files
    if (lstatSync(subPath).isDirectory()) {
      if (ignore(excludePatterns, relativePath, fileURLToPath(subPath))) {
        continue;
      }
      result.push(...await lintFolder(subPath, errorExpected, printMessages, excludePatterns, projectParser));
    } else if (minimatch(basename(relativePath), vhdlGlob)) {
      if (ignore(excludePatterns, relativePath, fileURLToPath(subPath))) {
        continue;
      }
      const text = readFileSyncNorm(subPath, { encoding: 'utf8' });
      const vhdlLinter = new VhdlLinter(subPath, text, projectParser, await projectParser.getDocumentSettings(subPath));
      if (vhdlLinter.parsedSuccessfully) {
        await vhdlLinter.checkAll();
      }
      if (errorExpected === false) {
        if (vhdlLinter.messages.length > 0) {
          const newMessage = {
            file: fileURLToPath(subPath),
            messages: vhdlLinter.messages
          };
          result.push(newMessage);
          if (printMessages) {
            console.log(prettyPrintMessages(rootPath, [newMessage]));
          }
        }
      } else {
        if (vhdlLinter.messages.length !== 1) {
          const newMessage: MessageWrapper = {
            file: fileURLToPath(subPath),
            messages: [
              ...vhdlLinter.messages,
              {
                message: `One message expected found ${vhdlLinter.messages.length}`,
                range: new OIRange(vhdlLinter.file, 0, 0)
              }]
          };
          result.push(newMessage);
          if (printMessages) {
            console.log(prettyPrintMessages(rootPath, [newMessage]));
          }
        }
      }
    }
  }

  return result;
}