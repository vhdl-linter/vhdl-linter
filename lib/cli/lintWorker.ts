import { lstatSync } from "fs";
import { parentPort, workerData } from "worker_threads";
import { ProjectParser } from "../projectParser";
import { VhdlLinter } from "../vhdlLinter";
import { readFileSyncNorm } from "./readFileSyncNorm";
import { MessageWrapper, getCodeClimate, prettyPrintMessages, readDirPath } from "./cliUtil";
import { OIRange } from "../parser/objects";
async function run_test(path: URL, errorExpected: boolean, printMessages: boolean, projectParser?: ProjectParser): Promise<MessageWrapper[]> {
  const result: MessageWrapper[] = [];
  if (!projectParser) {
    projectParser = await ProjectParser.create([path]);
  }
  for (const subPath of readDirPath(path)) {
    if (lstatSync(subPath).isDirectory()) {
      result.push(...await run_test(subPath, errorExpected, printMessages, projectParser));
    } else if (subPath.pathname.match(/\.vhdl?$/i)) {
      const text = readFileSyncNorm(subPath, { encoding: 'utf8' });
      const vhdlLinter = new VhdlLinter(subPath, text, projectParser, await projectParser.getDocumentSettings(subPath));
      if (vhdlLinter.parsedSuccessfully) {
        await vhdlLinter.checkAll();
      }
      if (errorExpected === false) {
        if (vhdlLinter.messages.length > 0) {
          const newMessage = {
            file: subPath.pathname,
            messages: vhdlLinter.messages
          };
          result.push(newMessage);
          if (printMessages) {
            console.log(prettyPrintMessages([newMessage]));
          }
        }
      } else {
        if (vhdlLinter.messages.length !== 1) {
          const newMessage: MessageWrapper = {
            file: subPath.pathname,
            messages: [
              ...vhdlLinter.messages,
              {
                message: `One message expected found ${vhdlLinter.messages.length}`,
                range: new OIRange(vhdlLinter.file, 0, 0)
              }]
          };
          result.push(newMessage);
          if (printMessages) {
            console.log(prettyPrintMessages([newMessage]));
          }
        }
      }
    }
  }

  return result;
}
(async () => {
  const { path, errorExpected, outputCodeClimate } = workerData as { path: string, errorExpected: boolean, outputCodeClimate: boolean };
  const messages = await run_test(new URL(path), errorExpected, outputCodeClimate === false);
  if (outputCodeClimate === true) {
    parentPort?.postMessage(getCodeClimate(messages));
  } else {
    parentPort?.postMessage(messages);
  }
})().catch(err => console.error(err));