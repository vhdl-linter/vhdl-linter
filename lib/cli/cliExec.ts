import { Command, InvalidArgumentError } from '@commander-js/extra-typings';
import { existsSync, lstatSync } from 'fs';
import { isAbsolute, join } from 'path';
import { cwd } from 'process';
import { fileURLToPath, pathToFileURL } from 'url';
import { DiagnosticSeverity } from 'vscode-languageserver';
import { ProjectParser } from '../projectParser';
import { OIDiagnostic } from '../vhdlLinter';
import { getCodeClimate, prettyPrintMessages } from './cliUtil';
import { lintFolder } from './lintFolder';

export async function cli(argv: string[]) {
  const integerParam = (v: string) => {
    const n = Number(v);
    if (Number.isInteger(n)) {
      return n;
    }
    throw new InvalidArgumentError(`Not an integer`);
  };
  const folderParam = (v: string) => {
    const path = isAbsolute(v) ? v : join(cwd(), v);
    if (existsSync(path) === false) {
      throw new InvalidArgumentError("Does not exist");
    }
    if (lstatSync(path).isDirectory() === false) {
      throw new InvalidArgumentError("Is not a directory");
    }
    return pathToFileURL(path);
  };
  const program = new Command('vhdl-linter')
    .description('A typescript based linter for vhdl')
    .argument('<folder>', 'The folder to lint', folderParam)
    .option('-j, --output-json', 'Output message in json compatible with Code Climate Engine Specification (Gitlab & Github compatible)', false)
    .option('-e, --exclude <pattern...>', 'Exclude pattern for linting. Use the ignore setting from the `vhdl-linter.yml` file to not parse files at all.', [])
    .option('-i, --max-info <count>', 'Number of information messages to trigger a nonzero exit code', integerParam, -1)
    .option('-w, --max-warnings <count>', 'Number of information messages to trigger a nonzero exit code', integerParam, -1)
    .parse(argv);
  const options = program.opts();
  const start = new Date().getTime();
  const folder = program.processedArgs[0];
  const projectParser = await ProjectParser.create([folder]);
  const filesWithMessages = await lintFolder(folder, false, options.outputJson === false, options.exclude, projectParser);
  await projectParser.stop();
  const allMessages = filesWithMessages.flatMap(m => m.messages);
  const warningCount = allMessages.filter(m => m?.severity === DiagnosticSeverity.Warning).length;
  const infoCount = allMessages.filter(m => m?.severity === DiagnosticSeverity.Information).length;
  // default is error
  function isError(m: OIDiagnostic) {
    return m?.severity !== DiagnosticSeverity.Warning && m?.severity !== DiagnosticSeverity.Information;
  }
  const errorCount = allMessages.filter(isError).length;
  if (options.outputJson) {
    const rootPath = fileURLToPath(projectParser.workspaces[0]!);
    console.log(JSON.stringify(getCodeClimate(filesWithMessages, rootPath), undefined, 2));
    const filesWithErrors = filesWithMessages.map(fileWithMessages => ({ ...fileWithMessages, messages: fileWithMessages.messages.filter(isError) })).filter(fileWithMessages => fileWithMessages.messages.length > 0);
    console.error(prettyPrintMessages(rootPath, filesWithErrors));
  } else {
    const timeTaken = new Date().getTime() - start;
    console.log(`Linted in ${(timeTaken / 1000).toFixed(2)}s:`);
    console.log(`${errorCount} error(s), ${warningCount} warning(s), ${infoCount} info(s)`);
  }
  if (errorCount > 0) {
    return 1;
  } else if (warningCount > options.maxWarnings && options.maxWarnings !== -1) {
    return 1;
  } else if (infoCount > options.maxInfo && options.maxInfo !== -1) {
    return 1;
  }
  return 0;
}
