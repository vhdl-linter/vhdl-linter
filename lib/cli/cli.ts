#!/usr/bin/env node
import { cli } from './cliExec';

(async () => {
  const returnCode = await cli(process.argv);
  process.exit(returnCode);
})().catch(err => console.error(err));
