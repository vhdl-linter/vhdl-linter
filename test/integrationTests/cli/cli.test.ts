import { expect, test } from "@jest/globals";
import { spawnSync } from "child_process";
import { join, sep } from "path";
import { cwd } from "process";
import { CodeClimateIssue } from "../../../lib/cli/cliUtil";

const testPath = join(__dirname, 'testFolder');
const cliPath = join(__dirname, '..', '..', '..', 'lib', 'cli', 'cli.ts');
function makeRelative(path: string) {
  return path.replace(cwd() + sep, '');
}

test('no parameters', () => {
  const process = spawnSync('npx', ['ts-node', cliPath, testPath], { encoding: 'utf8' });
  expect(process.status).toBe(3);
  expect(process.stdout.replace(/Linted in [\d.]+s:/, '')).toMatchSnapshot();
});

test('relative path', () => {
  const process = spawnSync('npx', ['ts-node', cliPath, makeRelative(testPath)], { encoding: 'utf8' });
  expect(process.status).toBe(3);
  expect(process.stdout.replace(/Linted in [\d.]+s:/, '')).toMatchSnapshot();
});


test('json', () => {
  const process = spawnSync('npx', ['ts-node', cliPath, testPath, '-j'], { encoding: 'utf8' });
  expect(process.status).toBe(3);
  const result = JSON.parse(process.stdout) as CodeClimateIssue[];
  expect(result).toHaveLength(3);
  const maskedResult = result.map(entry => ({ ...entry, location: { ...entry.location, path: makeRelative(entry.location.path) }, fingerprint: 'fake' }));
  expect(maskedResult).toMatchSnapshot();
});


test.each([
  ['**/error.vhd', 2], // file explicitly
  ['**/*.vhd', 0], // file wildcards
  ['subfolder', 0], // folders
  ['subfolder/info*', 2], // files in folder
])('exclude %s (expect %s messages)', (pattern: string, messageCount: number) => {
  const process = spawnSync('npx', ['ts-node', cliPath, testPath, '-e', pattern], { encoding: 'utf8' });
  expect(process.status).toBe(messageCount);
  expect(process.stdout.replace(/Linted in [\d.]+s:/, '')).toMatchSnapshot();
});

test('multiple excludes', () => {
  const process = spawnSync('npx', ['ts-node', cliPath, testPath, '-e', '**/info*', '**/error*'], { encoding: 'utf8' });
  expect(process.status).toBe(1);
  expect(process.stdout.replace(/Linted in [\d.]+s:/, '')).toMatchSnapshot();
});