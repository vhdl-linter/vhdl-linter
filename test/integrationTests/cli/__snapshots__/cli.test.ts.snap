// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exclude **/*.vhd (expect 0 messages) 1`] = `
"
0 error(s), 0 warning(s), 0 info(s)
"
`;

exports[`exclude **/error.vhd (expect 2 messages) 1`] = `
"subfolder/info.vhd:5 (r: 4:14 - 4:23)
  [34mPort is using resolved subtype (std_logic) should use unresolved type std_ulogic  (type-resolved)[0m
subfolder/warning.vhd:3 (r: 2:4 - 2:9)
  [33mNot reading input port 'input' (unused)[0m

0 error(s), 1 warning(s), 1 info(s)
"
`;

exports[`exclude subfolder (expect 0 messages) 1`] = `
"
0 error(s), 0 warning(s), 0 info(s)
"
`;

exports[`exclude subfolder/info* (expect 2 messages) 1`] = `
"subfolder/error.vhd:2 (r: 1:15 - 1:25)
  [31mobject 'std_ulogic' is referenced but not declared (not-declared)[0m
subfolder/warning.vhd:3 (r: 2:4 - 2:9)
  [33mNot reading input port 'input' (unused)[0m

1 error(s), 1 warning(s), 0 info(s)
"
`;

exports[`json 1`] = `
[
  {
    "categories": [
      "Style",
    ],
    "check_name": "vhdl-linter",
    "description": "object 'std_ulogic' is referenced but not declared (not-declared)",
    "fingerprint": "fake",
    "location": {
      "path": "test/integrationTests/cli/testFolder/subfolder/error.vhd",
      "positions": {
        "begin": {
          "column": 15,
          "line": 1,
        },
        "end": {
          "column": 25,
          "line": 1,
        },
      },
    },
    "severity": "blocker",
    "type": "issue",
  },
  {
    "categories": [
      "Style",
    ],
    "check_name": "vhdl-linter",
    "description": "Port is using resolved subtype (std_logic) should use unresolved type std_ulogic  (type-resolved)",
    "fingerprint": "fake",
    "location": {
      "path": "test/integrationTests/cli/testFolder/subfolder/info.vhd",
      "positions": {
        "begin": {
          "column": 14,
          "line": 4,
        },
        "end": {
          "column": 23,
          "line": 4,
        },
      },
    },
    "severity": "minor",
    "type": "issue",
  },
  {
    "categories": [
      "Style",
    ],
    "check_name": "vhdl-linter",
    "description": "Not reading input port 'input' (unused)",
    "fingerprint": "fake",
    "location": {
      "path": "test/integrationTests/cli/testFolder/subfolder/warning.vhd",
      "positions": {
        "begin": {
          "column": 4,
          "line": 2,
        },
        "end": {
          "column": 9,
          "line": 2,
        },
      },
    },
    "severity": "major",
    "type": "issue",
  },
]
`;

exports[`multiple excludes 1`] = `
"subfolder/warning.vhd:3 (r: 2:4 - 2:9)
  [33mNot reading input port 'input' (unused)[0m

0 error(s), 1 warning(s), 0 info(s)
"
`;

exports[`no parameters 1`] = `
"subfolder/error.vhd:2 (r: 1:15 - 1:25)
  [31mobject 'std_ulogic' is referenced but not declared (not-declared)[0m
subfolder/info.vhd:5 (r: 4:14 - 4:23)
  [34mPort is using resolved subtype (std_logic) should use unresolved type std_ulogic  (type-resolved)[0m
subfolder/warning.vhd:3 (r: 2:4 - 2:9)
  [33mNot reading input port 'input' (unused)[0m

1 error(s), 1 warning(s), 1 info(s)
"
`;

exports[`relative path 1`] = `
"subfolder/error.vhd:2 (r: 1:15 - 1:25)
  [31mobject 'std_ulogic' is referenced but not declared (not-declared)[0m
subfolder/info.vhd:5 (r: 4:14 - 4:23)
  [34mPort is using resolved subtype (std_logic) should use unresolved type std_ulogic  (type-resolved)[0m
subfolder/warning.vhd:3 (r: 2:4 - 2:9)
  [33mNot reading input port 'input' (unused)[0m

1 error(s), 1 warning(s), 1 info(s)
"
`;
